### PR TITLE
Add support for multi-directory globs

### LINKED ISSUE #49290: file/(hdfs?) globs allow to have patterns across different directories
**Use case**

We want to read only specific files in each directory, and do not have access to other files.

**Describe the solution you'd like**

```
SELECT
    *,
    _path,
    _file
FROM file('{a/1,b/2}.csv', CSV)


0 rows in set. Elapsed: 0.050 sec.

Received exception:
Code: 636. DB::Exception: Cannot extract table structure from CSV format file, because there are no files with provided path. You must specify table structure manually. (CANNOT_EXTRACT_TABLE_STRUCTURE)

SELECT
    *,
    _path,
    _file
FROM file('a/1.csv', CSV)

┌─c1─┬─c2─┬─_path────────────────┬─_file─┐
│  1 │  2 │ /home/xxxxxxx/a/1.csv │ 1.csv │
└────┴────┴──────────────────────┴───────┘

SELECT
    *,
    _path,
    _file
FROM file('b/2.csv', CSV)

┌─c1─┬─c2─┬─_path────────────────┬─_file─┐
│  3 │  4 │ /home/xxxxxxx/b/2.csv │ 2.csv │
└────┴────┴──────────────────────┴───────┘
```

It's already working in such way for s3(because there is no real directories here, but still)

So you can write something like:

```
SELECT     *, _path, _file
FROM s3('https://clickhouse-public-datasets.s3.amazonaws.com/my-test-bucket-768/{some/some_file_1,another/another_file_3}.csv',   NOSIGN, CSVWithNames);
```

Related https://github.com/ClickHouse/ClickHouse/issues/16682

### PR DESCRIPTION
**Pull request by zvonand**
Fixes #49290. Adds support to using `{str1, ...}` globs across different directories, e.g:
The following query will read both from `a/1.csv` and `b/2.csv`
```
SELECT
    *,
    _path,
    _file
FROM file('{a/1,b/2}.csv', CSV)
``` 

### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Add support for multi-directory selection (`{}`) globs.

### CODE REVIEW

File: src/Storages/StorageFile.cpp [hunk 1]
```diff
@@ -90,6 +90,61 @@ namespace ErrorCodes
...
+        if (re2::RE2::FullMatch(processed_suffix + dir_or_file_name, matcher))
+        {
+            if (next_slash_after_glob_pos == std::string::npos)
+            {
+                total_bytes_to_read += it->file_size();
+                result.push_back(it->path().string());
+            }
+            else
+            {
+                listFilesWithRegexpMatchingImpl(fs::path(full_path).append(processed_suffix).append(it->path().string()) / "" ,
```
Comment (Enmk) [inline src/Storages/StorageFile.cpp:133]: why not just
```C++
fs::path(full_path) / processed_suffix / it->path() / ""
```

And why do you need an empty string at the end?

File: src/Storages/HDFS/StorageHDFS.cpp [hunk 2]
```diff
@@ -63,23 +63,124 @@ namespace ErrorCodes
 }
 namespace
 {
+    Strings LSWithRegexpMatching(const String & path_for_ls,
+                                 const HDFSFSPtr & fs,
+                                 const String & for_match,
+                                 std::unordered_map<String, time_t> * last_mod_times);
+
+    /// When `{...}` has any `/`s, it must be processed in a different way
```
Comment (Enmk) [inline src/Storages/HDFS/StorageHDFS.cpp:71]: Please extend your commend with an explanation on this "different way" with something like a simplified algorithm in pseudocode.

**Comment by robot-ch-test-poll3** [comment]
This is an automated comment for commit cd1a3916a6ea755b24b475983f6f67447cebdd6a with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50559/cd1a3916a6ea755b24b475983f6f67447cebdd6a/ci_running.html)
The overall status of the commit is 🔴 failure

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>Style Check</td><td>Runs a set of checks to keep the code style clean. If some of tests failed, see the related log from the report</td><td>🔴 failure</td></tr>
</table>

**Comment by zvonand** [comment]
AST fails probably due to #50451

**Comment by zvonand** [comment]
Fails look unrelated

File: tests/queries/0_stateless/02771_complex_globs_in_storage_file_path.sql [hunk 3]
```diff
@@ -0,0 +1,10 @@
+-- Tags: no-replicated-database, no-parallel
+
+INSERT INTO TABLE FUNCTION file('02771/dir1/subdir11/data1.csv', 'CSV', 's String') SELECT 'This is file data1' SETTINGS engine_file_truncate_on_insert=1;
+INSERT INTO TABLE FUNCTION file('02771/dir2/subdir22/data2.csv', 'CSV', 's String') SELECT 'This is file data2' SETTINGS engine_file_truncate_on_insert=1;
+
+SELECT *, _file FROM file('02771/dir{?/subdir?1/da,2/subdir2?/da}ta1.csv', CSV);
+SELECT *, _file FROM file('02771/dir{?/subdir?1/da,2/subdir2?/da}ta2.csv', CSV);
+
+SELECT *, _file FROM file('02771/dir?/{subdir?1/data1,subdir2?/data2}.csv', CSV) WHERE _file == 'data1.csv';
+SELECT *, _file FROM file('02771/dir?/{subdir?1/data1,subdir2?/data2}.csv', CSV) WHERE _file == 'data2.csv';
```
Comment (SmitaRKulkarni) [inline tests/queries/0_stateless/02771_complex_globs_in_storage_file_path.sql:10]: Would be good to add a test with file path that does not exist

File: src/Storages/StorageFile.cpp [hunk 4]
```diff
@@ -92,6 +92,66 @@ namespace ErrorCodes
...
+ * Basically, a path with globs is processed by listFilesWithRegexpMatching. In case it detects multi-dir glob {.../..., .../...},
+ * listFilesWithFoldedRegexpMatching is in charge from now on.
+ * It works a bit different: it still recursively goes through subdirectories, but does not match every directory to glob.
+ * Instead, it goes many levels down (until the approximate max_depth is reached) and compares this multi-dir path to a glob.
+ * StorageHDFS.cpp has the same logic.
+*/
+void listFilesWithFoldedRegexpMatchingImpl(const std::string & path_for_ls,
+                                           const std::string & processed_suffix,
+                                           const std::string & suffix_with_globs,
+                                           const std::string & current_glob,
```
Comment (SmitaRKulkarni) [inline src/Storages/StorageFile.cpp:114]: not clear how `current_glob` is used

File: src/Storages/StorageFile.cpp [hunk 5]
```diff
@@ -92,6 +92,66 @@ namespace ErrorCodes
...
+    const std::string & path_for_ls,
+    const std::string & for_match,
+    size_t & total_bytes_to_read,
+    std::vector<std::string> & result,
+    bool recursive = false);
+
+/*
+ * When `{...}` has any `/`s, it must be processed in a different way:
+ * Basically, a path with globs is processed by listFilesWithRegexpMatching. In case it detects multi-dir glob {.../..., .../...},
+ * listFilesWithFoldedRegexpMatching is in charge from now on.
```
Comment (SmitaRKulkarni) [inline src/Storages/StorageFile.cpp:106]:
```suggestion
 * listFilesWithFoldedRegexpMatchingImpl is in charge from now on.
```

**Review by SmitaRKulkarni — COMMENTED** [review]
Rest all LGTM

**Comment by zvonand** [comment]
Stateless tests (debug, s3 storage) [1/6]
* * `02479_mysql_connect_to_self` -- #50911 
* `02697_stop_reading_on_first_cancel`, `02466_distributed_query_profiler`, `02228_merge_tree_insert_memory_usage` -- look unrelated to me

Stateless tests (release, DatabaseReplicated) [2/4], Stateless tests (release, wide parts enabled):
* `01193_metadata_loading` -- fails from time to time in master (e.g. [959f66eb8a7611101c98d69dfbae61eda4c7476d](https://s3.amazonaws.com/clickhouse-test-reports/0/959f66eb8a7611101c98d69dfbae61eda4c7476d/stateless_tests__aarch64_.html)

File: src/Storages/StorageFile.cpp (hunk 4 above)
Comment (zvonand) [reply src/Storages/StorageFile.cpp]: it remained from an older version. removed this parameter from both files

File: tests/queries/0_stateless/02771_complex_globs_in_storage_file_path.sql (hunk 3 above)
Comment (zvonand) [reply tests/queries/0_stateless/02771_complex_globs_in_storage_file_path.sql:12]: added

**Comment by zvonand** [comment]
@SmitaRKulkarni 
Updated as per your review. Test fails analysis is in the comment above: nothing related to this PR there

**Comment by SmitaRKulkarni** [comment]
@zvonand : Could you please add a test with file path that does not exist for hdfs as well ?

**Review by SmitaRKulkarni — APPROVED** [review]

**Comment by alexey-milovidov** [comment]
Please do not merge without fixing the tests.

**Comment by zvonand** [comment]
Stateless tests (release, analyzer):
* `00992_system_parts_race_condition_zookeeper_long` -- flaky, #49400, #50600. Unrelated.


Integration tests (tsan) [3/6]:
* whole bunch of `test_storage_azure_blob_storage` -- already failing in PRs that are also unrelated to it: see [checks DB](https://play.clickhouse.com/play?user=play#c2VsZWN0IHB1bGxfcmVxdWVzdF9udW1iZXIgYXMgcHIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMgYXMgc3RhdHVzLCByZXBvcnRfdXJsCmZyb20gY2hlY2tzCndoZXJlIHRvRGF0ZShjaGVja19zdGFydF90aW1lKSA+ICcyMDIzLTA2LTEzJyBhbmQgdGVzdF9uYW1lIGxpa2UgJyV0ZXN0X3N0b3JhZ2VfYXp1cmVfYmxvYl9zdG9yYWdlJScgYW5kIHN0YXR1cyA9PSAnRkFJTCcKb3JkZXIgYnkgY2hlY2tfc3RhcnRfdGltZSBkZXNj). Unrelated here.


Integration tests (asan) [4/6]:
* test_multiple_disks/test.py::test_start_stop_moves[replicated] -- also [fails](https://play.clickhouse.com/play?user=play#c2VsZWN0IHB1bGxfcmVxdWVzdF9udW1iZXIgYXMgcHIsIGNoZWNrX3N0YXJ0X3RpbWUsIGNoZWNrX25hbWUsIHRlc3RfbmFtZSwgdGVzdF9zdGF0dXMgYXMgc3RhdHVzLCByZXBvcnRfdXJsCmZyb20gY2hlY2tzCndoZXJlIHRvRGF0ZShjaGVja19zdGFydF90aW1lKSA+ICcyMDIzLTA2LTEzJyBhbmQgdGVzdF9uYW1lIGxpa2UgJyV0ZXN0X211bHRpcGxlX2Rpc2tzL3Rlc3QucHk6OnRlc3Rfc3RhcnRfc3RvcF9tb3ZlcyUnIGFuZCBzdGF0dXMgPT0gJ0ZBSUwnCm9yZGVyIGJ5IGNoZWNrX3N0YXJ0X3RpbWUgZGVzYw==) from time to time in neighbor PRs and in master. Unrelated.

**Comment by zvonand** [comment]
@alexey-milovidov which of the fails above do you think are relative to these changes?

**Comment by alexey-milovidov** [comment]
It does not matter if the tests are related or not.
Failed tests means no merge.

**Comment by zvonand** [comment]
Integration tests (asan) [1/6]:
* `test_zookeeper_config/test.py` -- #51464

Integration tests (asan) [4/6]:
* `test_multiple_disks/test.py::test_start_stop_moves[replicated]` -- #45404, 

Integration tests (release) [3/4] :
* test_host_regexp_multiple_ptr_records/test.py::test_host_regexp_multiple_ptr_v4 -- #51471

Integration tests (release) [4/4]:
* `test_ssl_cert_authentication/test.py::test_https_non_ssl_auth` -- #49400

Stateless tests (tsan, s3 storage) [4/5]:
* 01945_system_warnings -- probably fixed by #51453

Stateless tests (ubsan) [2/2]:
* 00992_system_parts_race_condition_zookeeper_long -- #50434

**Comment by zvonand** [comment]
**Integration tests (release) [4/4]:**
* test_https_wrong_cert -- #49400

**Stateless tests (release, s3 storage) [1/2]:**
* 01415_sticking_mutations -- #51538

**Comment by zvonand** [comment]
The only fail is `00834_kill_mutation`: #51818

**Review by alexey-milovidov — APPROVED** [review]

**Pull request merged**